documentation/writing_tests.md

```diff
@@ -0,0 +1,1473 @@
+- And also in how we use these by looking at any mention of `rollup` on the [template project's package.json](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/template/template/package.json).
+
+### Karma
+
+Karma is a tool that allows us to run the same TypeScript tests we run on node, but on the browsers. The Karma runner as a tool has a considerably low barrier of entry, but it is also pretty extensive. We recommend going through the following resources to learn more about Karma:
+
+- About Karma's [configuration file](http://karma-runner.github.io/4.0/config/configuration-file.html).
+- About Karma's handling of [files](http://karma-runner.github.io/4.0/config/files.html).
+- About Karma's support for [browsers](http://karma-runner.github.io/4.0/config/browsers.html).
+- About Karma's [preprocessors](http://karma-runner.github.io/4.0/config/preprocessors.html).
+- About Karma's [plugins](http://karma-runner.github.io/4.0/config/plugins.html).
+
+#### Karma in our dependencies
+
+If you want to add `karma` as a dev dependency to a new project, please use [Rush](https://rushjs.io/) since we're most likely using the same version of Karma in more projects. First make sure you are in the root folder of that project, then you can run the following command:
+
+```
+rush add --dev -p karma
+```
+
+To fulfill our needs, we use Karma with some plugins. They're the following:
+
+- [`karma-coverage`](https://www.npmjs.com/package/karma-coverage):
+  Generate code coverage using [Istanbul](https://github.com/gotwarlost/istanbul).
+- [`karma-env-preprocessor`](https://www.npmjs.com/package/karma-env-preprocessor):
+  It's a Karma preprocessor which makes environment variables available to our tests.
+- [`karma-json-preprocessor`](https://www.npmjs.com/package/karma-json-preprocessor):
+  It's a Karma preprocessor for converting JSON files into JS variables.
+- [`karma-json-to-file-reporter`](https://www.npmjs.com/package/karma-json-to-file-reporter):
+  It's a Karma reporter that save JSON messages from log to file. We use this to create real local files while executing tests in the browser. It's specifically used by [the recorder](#the-recorder).
+- [`karma-junit-reporter`](https://www.npmjs.com/package/karma-junit-reporter):
+  It's a Karma reporter for the [JUnit XML format](https://llg.cubic.org/docs/junit/).
+- [`karma-mocha`](https://www.npmjs.com/package/karma-mocha):
+  It's an adapter for the [Mocha](http://mochajs.org/) testing framework.
+- [`karma-mocha-reporter`](https://www.npmjs.com/package/karma-mocha-reporter):
+  It's a Karma reporter plugin with mocha style logging.
+- [`karma-remap-istanbul`](https://www.npmjs.com/package/karma-remap-istanbul):
+  A Karma reporter that enables remapped reports on watch.
+- [`karma-chrome-launcher`](https://www.npmjs.com/package/karma-chrome-launcher):
+  Launcher for Google Chrome, Google Chrome Canary and Google Chromium.
```

Comment:
Are you not adding the launchers for other browsers?
`karma-firefox-launcher`, `karma-ie-launcher`, `karma-edge-launcher`

Reply:
We're not using these anywhere, or are we? I am at least not aware.

Reply:
These are present in all the storage packages, we do manual testing in these browsers right before the release.

Reply:
I believe we should document this. Have we made instructions for this anywhere?

Reply:
Storage packages have a requirement to support IE which might not be the case for other services.
Along with IE, firefox and edge are tested too before the release(part of the release checklist for storage libraries).

Maybe you can mention the existence of these karma-launchers and let developers decide if they want to add them for their service.